src/promql/src/functions/idelta.rs

```diff
@@ -0,0 +1,243 @@
+    // time index column and value column
+    fn input_type() -> Vec<DataType> {
+        vec![
+            RangeArray::convert_data_type(DataType::Int64),
+            RangeArray::convert_data_type(DataType::Float64),
+        ]
+    }
+
+    fn return_type() -> DataType {
+        DataType::Float64
+    }
+
+    fn calc(input: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionError> {
+        // construct matrix from input
+        assert_eq!(input.len(), 2);
+        let ts_array = extract_array(&input[0])?;
+        let value_array = extract_array(&input[1])?;
+
+        let ts_range: RangeArray = RangeArray::try_new(ts_array.data().clone().into())?;
+        let value_range: RangeArray = RangeArray::try_new(value_array.data().clone().into())?;
+        error::ensure(
+            ts_range.len() == value_range.len(),
+            DataFusionError::Execution(format!(
+                "input arrays should have the same length, found {} and {}",
+                ts_range.len(),
+                value_range.len()
+            )),
+        )?;
+        error::ensure(
+            ts_range.value_type() == DataType::Int64,
+            DataFusionError::Execution(format!(
+                "expect Int64 as time index array's type, found {} ",
+                ts_range.value_type()
+            )),
+        )?;
+        error::ensure(
+            value_range.value_type() == DataType::Float64,
+            DataFusionError::Execution(format!(
+                "expect Int64 as time index array's type, found {} ",
+                value_range.value_type()
+            )),
+        )?;
+
+        // calculation
+        let mut result_array = Vec::with_capacity(ts_range.len());
+
+        for index in 0..ts_range.len() {
+            let timestamps = ts_range.get(index).unwrap();
+            let timestamps = timestamps
+                .as_any()
+                .downcast_ref::<Int64Array>()
+                .unwrap()
+                .values();
+
+            let values = value_range.get(index).unwrap();
+            let values = values
+                .as_any()
+                .downcast_ref::<Float64Array>()
+                .unwrap()
+                .values();
+            error::ensure(
+                timestamps.len() == values.len(),
+                DataFusionError::Execution(format!(
+                    "input arrays should have the same length, found {} and {}",
+                    timestamps.len(),
+                    values.len()
+                )),
+            )?;
+
+            let len = timestamps.len();
+            if len < 2 {
+                result_array.push(0.0);
+                continue;
+            }
+
+            // if is delta
+            if !IS_RATE {
+                result_array.push(values[len - 1] - values[len - 2]);
+                continue;
+            }
+
+            // else is rate
+            let sampled_interval = (timestamps[len - 1] - timestamps[len - 2]) / 1000;
```

Comment:
Can we add a comment here? Do we consume the timestamp unit is milliseconds here?

Reply:
I add a todo comment. I've not figured out how to adapt other timestamp precisions to those functions. Plan to investigate this later after more functions are implemented.